data/transactions/logic/assembler.go

```diff
@@ -1184,17 +1150,177 @@ func typeLoads(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes) {
 	return nil, StackTypes{scratchType}
 }
 
-// keywords or "pseudo-ops" handle parsing and assembling special asm language
-// constructs like 'addr' We use an OpSpec here, but it's somewhat degenerate,
-// since they don't have opcodes or eval functions. But it does need a lot of
-// OpSpec, in order to support assembly - Mode, typing info, etc.
-var keywords = map[string]OpSpec{
-	"int":  {0, "int", nil, proto(":i"), 1, assembler(asmInt)},
-	"byte": {0, "byte", nil, proto(":b"), 1, assembler(asmByte)},
+func pseudoImmediatesError(ops *OpStream, name string, specs map[int]OpSpec) {
+	immediateCounts := make([]int, len(specs))
+	i := 0
+	for numImms := range specs {
+		immediateCounts[i] = numImms
+		i++
+	}
+	sort.SliceStable(immediateCounts, func(j, k int) bool { return immediateCounts[j] < immediateCounts[k] })
+	errMsg := fmt.Sprintf("%s expects", name)
+	for i, numImms := range immediateCounts {
+		if i+1 < len(immediateCounts) {
+			errMsg += fmt.Sprintf(" %d or", numImms)
+		} else {
+			errMsg += fmt.Sprintf(" %d", numImms)
+		}
+	}
+	errMsg += " immediate arguments"
+	ops.error(errMsg)
+}
+
+// getSpec finds the OpSpec we need during assembly based on it's name, our current version, and the immediates passed in
+// Note getSpec handles both normal OpSpecs and those supplied by versionedPseudoOps
+func getSpec(ops *OpStream, name string, args []string) (OpSpec, string, bool) {
+	pseudoSpecs, ok := ops.versionedPseudoOps[name]
+	if ok {
+		pseudo, ok := pseudoSpecs[len(args)]
+		if !ok {
+			// Could be that pseudoOp wants to handle immediates itself so check -1 key
+			pseudo, ok = pseudoSpecs[anyImmediates]
+			if !ok {
+				// Number of immediates supplied did not match any of the pseudoOps of the given name, so we try to construct a mock spec that can be used to track types
+				pseudoImmediatesError(ops, name, pseudoSpecs)
+				proto, version, ok := mergeProtos(pseudoSpecs)
+				if !ok {
+					return OpSpec{}, "", false
+				}
+				pseudo = OpSpec{Name: name, Proto: proto, Version: version, OpDetails: OpDetails{asm: func(*OpStream, *OpSpec, []string) error { return nil }}}
+			}
+		}
+		pseudo.Name = name
+		if pseudo.Version > ops.Version && (ops.Version != 0 || pseudo.Version != 1) {
```

Comment:
@iten-alg Requests for clarification:
* In what scenario can `ops.Version == 0`?  My interpretation is:  `ops.Version` is unconditionally initialized to a non-zero value prior to `getSpec` initialization.
* Can you remind me why `pseudo.Version ==  1` is a special case here?

Reply:
Good catch! I had thought that since we constructed a spec map for version 0, we would use that, but it does seem like we just default the version to 1 during assembly and only use the map in eval. (version 1 would have otherwise been a special case b/c it is essentially a v1 spec not in the v0 spec map since we don't add pseudo-ops to the maps)

On a side note, do you know the reason for the difference between eval and assembler in how they handle v0?

Reply:
I don't think v0 ever actually happened. I just try to leave that as untouched as possible.

Reply:
Update @jannotti @michaeldiamant : I believe we actually need the check b/c for some reason AssemblerNoVersion is actually uint64 max, not 0. No clue why we check if the version is uint64 max and then set the version to 0 if it is, but this does mean that assembly does in fact use the version 0 table if no version is given.

Reply:
@iten-alg Thanks for the research.  Agree to _not_ tamper with v0 and v1 artifacts - resolving.